### PR TITLE
Checking for open data

### PR DESCRIPTION
**Pull request by coverbeck**
**Description**
Changes the way we search for open data, by only looking at test values that correspond to the workflow's file inputs. Only done for WDL in this PR. CWL ticket: #5385 .

The previous approach was to look the values in a test parameter file, and for any values that were valid URLs as recognized by Java, see if they were open.

The new approach addresses this:

* If a file input value was a local, relative path, or a gs/s3 url, it would be skipped (because it's not  an URL or protocol that Java recognizes), and the test file would be considered open. There were lots of these of versions marked as open when they weren't (almost all). So many that I think it best to just reset.

**Review Instructions**
1. Add a test parameter file with open data for all file inputs (https urls you can download without auth) to a version
2. Get it onto Dockstore (.dockstore.yml or refresh, whichever you prefer)
3. With the version selected, verify that the Info tab contains: `Test Parameter File With Open Data: Yes`
4. Also verify that the text is not visible when positioned on another version.

**Issue**
#5339 

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by coverbeck** [comment]
Test failures are legit, need to update, but going to wait on feedback first.

**Review by svonworl — COMMENTED** [review]
The "New Approach" makes sense, seems like it should work fine.  Will do a detailed line-by-line review when you re-request.

**Comment by denis-yuen** [comment]
> * If a file input parameter value is not a valid URL, e.g., it's a local file reference, we skip it.

Is there a reason we skip it instead of failing it?

**Review by denis-yuen — COMMENTED** [review]
don't entirely understand the description, but the general approach (keep track of parameters and match them with open urls/data seems ok)

**Comment by coverbeck** [comment]
> > * if for some reason there is a URL parameter value that is not a file input, if it's not public, we could incorrectly flag the parameter file as not open.
> 
> Not sure I understand this one, can you elaborate?

It's a real unlikely edge case; let's say the WDL was:

```
workflow w { 
  input {
      String label = "The label" 
      File theFile
    }
...
```

and your test parameter file looked like this:
```
{
   "w.label": "https://urldoesnotexist.com",
   "w.theFile": "https://urlexists.com"
```

In the old approach, the parameter file would not be deemed open because of the value `https://urldoesnotexist.com` is not a valid url, even though the parameter `w.label` is a string.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java [hunk 1]
```diff
@@ -69,112 +63,47 @@ private static Optional<Boolean> checkUrl(String url, String baseURL) {
...
+            return false;
+        });
     }
 
     /**
-     * Get all the URLs from a JSON file.
-     *
-     * @param contents Contents of a JSON file
-     * @return The URLs
+     * Checks whether all <code>possibleUrls</code> are open access urls. Returns <code>Optional.of(true)</code>
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:87]: I should know this, but could use code comment. What makes a URL open access? Just that it doesn't require authentication?

**Review by denis-yuen — COMMENTED** [review]
some comments/questions for clarification

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java [hunk 2]
```diff
@@ -69,112 +63,47 @@ private static Optional<Boolean> checkUrl(String url, String baseURL) {
...
-
-    private static void getUrlsFromJSON(JSONObject jsonObject, Set<String> urls) {
-        Iterator<?> keys = jsonObject.keys();
-        while (keys.hasNext()) {
-            String key = (String) keys.next();
-            Object o = jsonObject.get(key);
-            getUrlsFromJSON(o, urls);
+    public static Optional<Boolean> checkUrls(Set<String> possibleUrls, String baseURL) {
+        if (possibleUrls.isEmpty() || hasMalformedUrl(possibleUrls)) {
+            // If there are no urls, or any values that aren't urls, it's not open data
```
[the remark this answers is not in the document]
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:96]: > Seems intentional, so asking the question. Wouldn't "no urls" be essentially open data? i.e. a workflow that runs with no parameters or requires no input

You could argue this both ways, and I could go either way. Should a "Hello world" workflow be considered open data? For now, I opted that it's not, that no data != open data.  On the other hand, if you're looking for workflows you can run without restricted access data, it would be useful.

Maybe we should come up with another term instead of open data, something that indicates runnable as is without any access controlled data. Trying to think of one.

[the remark this answers is not in the document]
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:96]: Ok with a third term, without a third term I'd argue that yes they should be open data. 
Imagine, yes, a hello world workflow, a workflow that only takes integer parameters, or something like a chatgpt container that only takes a String question

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java [hunk 3]
```diff
@@ -960,6 +957,17 @@ default Optional<String> convertMapsToContent(final String mainDescName, final T
...
     }
 
+    /**
+     * Indicates if the workflow version has a test parameter file with all file inputs values
+     * being open data.
+     * @param workflowVersion
+     * @param checkUrlLambdaUrl
+     * @return if unable to determine, Optional.empty(), otherwise an non-empty boolean
+     */
+    default Optional<Boolean> hasPublicAccessibleUrls(WorkflowVersion workflowVersion, final String checkUrlLambdaUrl) {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java:967]: Idea: instead of passing in the lambda checking url, which couples the ensuing code to that particular mechanism, we could create a `UrlChecker` interface, and create an implementation that uses the lambda, and pass that to `hasPublicAccessibleUrls`.  Something like:
```
interface UrlChecker {
    Optional<Boolean> checkUrls(Set<String> urls);
}

public class LambdaUrlChecker implements UrlChecker {
    private final String lambdaUrl;
    public LambdaUrlChecker(String lambdaUrl) {
        this.lambdaUrl = lambdaUrl;
    }
    Optional<Boolean> checkUrls(Set<String> urls) {
        return CheckUrlHelper.checkUrls(urls, lambdaUrl);
    }
}
```
That way, if we ever change the url checking mechanism, like move it to something other than the lambda, it's super easy, and the existing code just works.  We can also mock the url checking more easily.

In general, for future refactors, I recommend we do more of this kind of thing.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:96]: Brainstorming here, an idea would be to reframe the terminology, and use the term "open" instead of "open data".  A workflow is "open" if it doesn't require restricted resources...

[the remark this answers is not in the document]
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:96]: I don't think we can detect if a workflow requires restricted resources. Well, not today -- we'd have to scan the code and see if there are any hard-coded references to restricted resources. We can only detect if a test parameter file references open/restricted resources. Here are the scenarios:

Id | Use Case  | Notes  |
-- | ---------- | -------- |
1 | Workflow has no input parameters  |     |
2 | Workflow has input parameters, but none are files |  |
3 | Workflow has file input parameters and provides a test parameter file with open access data for every file input| |
4 | Workflow has file input parameters and has no test param file, or has no test param file without open access data for every file parameter | Note that many test param files have references to local data, i.e., relative file paths, that could be open data, but we have no way of knowing |

The PR currently only flags option 3 as open access. Denis wants cases 1 and 2 covered as well.

So, we want to be able to find workflows that, as registered on Dockstore, are completely runnable on their own as they are on Dockstore (except for Docker and the workflow engine), without any additional access required.

Brainstorm of ideas, none of which I like. I'm hoping they may trigger a good idea for somebody :). Keep in mind we'll need to surface this in the search facets.

* Runnable-as-is; Runnable locally; Runnable
* Self-contained
* Complete; Complete bundle

I'll try to think of something better over the weekend. :)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+     * if source file is not a test YAML or JSON, or if the file content has a syntax error.
+     * @param sourceFile
+     * @return
+     */
+    public static Optional<JSONObject> testFileAsJsonObject(SourceFile sourceFile) {
+        final TestFileType testFileType = findTestFileType(sourceFile);
+        try {
+            switch (testFileType) {
+            case YAML -> {
+                Yaml yaml = new Yaml(new SafeConstructor());
```
[the remark this answers is not in the document]
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java:74]: Interesting, we use this pattern in several places.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:96]: > The PR currently only flags option 3 as open access. Denis wants cases 1 and 2 covered as well.

:heavy_check_mark: 

I think we just use "open data" vs "closed data" and just have a :information_source:  icon or something explaining. I think people are kinda used to the zero or null case being a little weird.

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml [hunk 5]
```diff
@@ -294,4 +294,10 @@
             CREATE UNIQUE INDEX case_insensitive_workflow_workflowname on workflow using btree (sourcecontrol, organization, repository, lower(workflowname)) where workflowname is not null and id > 18876;
         </sql>
     </changeSet>
+    <changeSet id="reset open data" author="coverbeck">
+        <update tableName="version_metadata">
+            <!--  Not explicitly specifying a value sets it to null-->
+            <column name="publicaccessibletestparameterfile"></column>
```
Comment (coverbeck) [inline dockstore-webservice/src/main/resources/migrations.1.14.0.xml:300]: Ideally we would rename this column and the corresponding property as it now means "everything needed to run the workflow is on Dockstore", i.e., no file inputs, or file inputs and a test parameter file with all public access data. But renaming properties is a hassle if we want the UI to work with no downtime and the CLI not to break. So I'm inclined to just leave it.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java (hunk 3 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java]: Good idea, added.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            .filter(sf -> Objects.equals(sf.getPath(), workflowVersion.getWorkflowPath()))
+            .findFirst();
+    }
+
+    /**
+     * Finds all test files in a workflow version
+     * @param workflowVersion
+     * @return
+     */
+    public static List<SourceFile> findTestFiles(WorkflowVersion workflowVersion) {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java:58]: Would it make sense to move `findPrimaryDescriptor` and `findTestFiles` into `WorkflowVersion`?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java [hunk 7]
```diff
@@ -26,36 +26,30 @@
...
 import java.util.stream.Collectors;
 import javax.ws.rs.core.UriBuilder;
-import org.json.JSONArray;
-import org.json.JSONObject;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.yaml.snakeyaml.Yaml;
 
-public final class CheckUrlHelper {
+public final class CheckUrlHelper implements CheckUrlInterface {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlHelper.java:37]: Since we use the `Helper` suffix to denote helper classes, and this class is no longer a helper, but rather a specialization of `CheckUrlInterface`, suggest we use the typical naming scheme, i.e. `LambdaCheckUrl` or similar.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlInterface.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * <ul>
+ *     <li>Has a protocol recognized out of the box by Java. Note that this does NOT include S3
+ *     and GS URIs</li>
+ *     <li>Is not a local file; caveat to above, as file:// is a protocol recognized by Java</li>
+ *     <li>Is accessible without authentication</li>
+ * </ul>
+ *
+ */
+public interface CheckUrlInterface {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlInterface.java:33]: STRONG OPINION WARNING:
So... I'm gonna lobby against appending `Interface` to our interface names.  Ideally, most code will operate on/with interface references, and IMHO, the repeated `Interface` verbiage clouds the code and provides very little value.  We know it's an interface, the IDE knows it's an interface, no need to belabor the point. :)  Similar feelings about `Impl`. An example of `Interface` and `Impl`-free naming:

```
interface Foo
abstract class BaseFoo implements Foo
class UsefulFoo extends BaseFoo
class ExtremeFoo extends BaseFoo
class CompositeFoo implements Foo
class LoggingFoo implements Foo
```
And most everywhere that uses a `Foo` is passed a `Foo` reference, rather than a reference of an implementing class type.

**Comment by svonworl** [comment]
Might want to check the Sonarcloud, there's some borderline interesting stuff in there.

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml (hunk 5 above)
Comment (denis-yuen) [reply dockstore-webservice/src/main/resources/migrations.1.14.0.xml:300]: Not too worried if there is a good comment

File: dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java [hunk 9]
```diff
@@ -17,54 +17,26 @@
...
-        String s = Files.asCharSource(file, StandardCharsets.UTF_8).read();
-        assertTrue(CheckUrlHelper.checkTestParameterFile(s, "fakeBaseUrl", TestFileType.YAML).isEmpty());
-        assertTrue(CheckUrlHelper.checkTestParameterFile(s, "fakeBaseUrl", TestFileType.JSON).isEmpty());
+    void testCheckUrls() {
+        final CheckUrlHelper checkUrlHelper =
+            new CheckUrlHelper("https://this.url.is.unused.in.these.tests");
+        assertEquals(UrlStatus.ALL_OPEN, checkUrlHelper.checkUrls(Set.of()),
+            "No urls should return all open");
+        assertEquals(UrlStatus.NOT_ALL_OPEN, checkUrlHelper.checkUrls(Set.of("localfile.cram")), "A local file is a malformed url");
+        assertEquals(UrlStatus.NOT_ALL_OPEN, checkUrlHelper.checkUrls(Set.of("s3://human-pangenomics/working/HPRC_PLUS/HG005/raw_data/Illumina/child/5A1-24481579/5A1_S5_L001_R1_001.fastq.gz")),
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java:39]: Phase 2 suggestion, we do have vfs2 which should recognize s3. 
(actually no, requires library, phase 2 anyway)

File: dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java [hunk 10]
```diff
@@ -17,54 +17,26 @@
...
     @Test
-    void getUrlsFromInvalidFile() throws IOException {
-        File file = new File(ResourceHelpers.resourceFilePath("valid_description_example.wdl"));
-        String s = Files.asCharSource(file, StandardCharsets.UTF_8).read();
-        assertTrue(CheckUrlHelper.checkTestParameterFile(s, "fakeBaseUrl", TestFileType.YAML).isEmpty());
-        assertTrue(CheckUrlHelper.checkTestParameterFile(s, "fakeBaseUrl", TestFileType.JSON).isEmpty());
+    void testCheckUrls() {
+        final CheckUrlHelper checkUrlHelper =
+            new CheckUrlHelper("https://this.url.is.unused.in.these.tests");
+        assertEquals(UrlStatus.ALL_OPEN, checkUrlHelper.checkUrls(Set.of()),
```
Comment (denis-yuen) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java:36]: Code comment potentially, if I understand correctly (code comment to clarify if not), the lambda URL `https://this.url.is.unused.in.these.tests` is not being used because none of these URLs are legitimate http(s) urls. If they were the test would not work because we don't have a lambda spun up for the tests

**Review by denis-yuen — COMMENTED** [review]
Looks ok in general, some clean-up required as suggested

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlInterface.java (hunk 8 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CheckUrlInterface.java:33]: I prefer to not have `Interface` myself, but if my grep is correct, we have 12 public interfaces declared in the web service, 9 of whose names end with `Interface`. I'd rather stick to the existing pattern and change it globally on team consensus rather than doing a one-off here.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java (hunk 6 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java:58]: Hmm, perhaps

File: dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java (hunk 9 above)
Comment (coverbeck) [reply dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/CheckUrlHelperTest.java:39]: #5397

**Comment by coverbeck** [comment]
> Might want to check the Sonarcloud, there's some borderline interesting stuff in there.

Stupid SonarCloud, the 3 bugs it's highlighting are from code remediation code they suggested and I literally copied from their site. :(

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java (hunk 6 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceFileHelper.java:58]: I hesitated because our Hibernate entities are mainly just persistent properties with getters/setters, with a few exceptions.

But the code otherwise does seem to make more sense in WorkflowVersion, so I will move.

**Review by denis-yuen — APPROVED** [review]
Codacy and sonarcloud ignorable.

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/CheckUrlIT.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2021 OICR and UCSC
```
Comment (kathy-t) [inline dockstore-integration-testing/src/test/java/io/dockstore/client/cli/CheckUrlIT.java:2]: 2021 -> 2023

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/LambdaUrlChecker.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2021 OICR and UCSC
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/LambdaUrlChecker.java:2]: 2021 -> 2023

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java [hunk 13]
```diff
@@ -250,47 +256,29 @@ private WorkflowVersion updateDBVersionSourceFilesWithRemoteVersionSourceFiles(W
...
 
     /**
-     * Sets the publicly accessible URL version metadata.
-     * If at least one test parameter file is publicly accessible, then version metadata is true
-     * If there's 1+ test parameter file that is null but there's no false, then version metadata is null
-     * If there's 1+ test parameter file that is false, then version metadata is false
+     * Sets the publicly accessible URL version metadata. If at least one test parameter file is
+     * publicly accessible, then version metadata is true If there's 1+ test parameter file that is
+     * null but there's no false, then version metadata is null If there's 1+ test parameter file
+     * that is false, then version metadata is false
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:270]: Nit: I find this formatting a little hard to read. Maybe use a list?

File: dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/LambdaUrlCheckerTest.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2021 OICR and UCSC
```
Comment (kathy-t) [inline dockstore-webservice/src/test/java/io/dockstore/webservice/helpers/LambdaUrlCheckerTest.java:2]: 2021 -> 2023

**Review by kathy-t — APPROVED** [review]

**Review by svonworl — APPROVED** [review]

File: dockstore-integration-testing/src/test/java/io/dockstore/client/cli/CheckUrlIT.java (hunk 11 above)
Comment (coverbeck) [reply dockstore-integration-testing/src/test/java/io/dockstore/client/cli/CheckUrlIT.java:2]: Grr, was actually renamed from CheckUrlHelperIT, which had the old copyright. I guess the file changed enough that Git didn't auto track it as a rename and I didn't explicitly do a `git mv`. :( It changed so much, it's probably just as well.
